proto: introduce seastar friendly protogen

rockwotj opened the pull request:
This PR introduces a new protoc plugin that can generate protos that are "seastar friendly", meaning they are not prone to oversized allocations and have serializations that yield control of the CPU when there is IO pending. The plugin is written in `golang` and lives in `bazel/pbgen/main.go` using the offical protobuf library (which includes a nice reflect library to write plugins in).

There is also now a set of Bazel rules to do said generation of classes, and serde routines to convert protos into iobufs (and back).

Usage looks like this:

```bzl
load("@rules_proto//proto:defs.bzl", "proto_library")
load("//bazel/pbgen:pbgen.bzl", "redpanda_proto_library")

proto_library(
    name = "version_proto",
    srcs = ["version.proto"],
    visibility = ["//visibility:public"],
    deps = ["//proto/redpanda/pbgen:options_proto"],
)

redpanda_proto_library(
    name = "version_redpanda_proto",
    protos = [":version_proto"],
    visibility = ["//visibility:public"],
)
```

The `proto_library` is the default rules_proto way of compiling protos to descriptor sets, and `redpanda_proto_library` generates our seastar friendly code. Note that `redpanda_proto_library` requires you passing in other `redpanda_proto_library` rules that it depends on.

The API of a generated proto is fairly similar to the offical protobuf library outputs, for the following proto (dependencies omitted):

```proto
/**
 * Represents a person with various fields, nested messages, and enums.
 */
message Person {
  // Unique identifier for the person
  int32 id = 1;

  // Full name
  string name = 2;

  // Email address
  string email = 3;

  // List of phone numbers
  repeated PhoneNumber phones = 4;

  // Gender enum
  Gender gender = 5;

  // Address — oneof home or work
  oneof address {
    string home_address = 6;
    string work_address = 7;
  }

  // Metadata (string → string map)
  map<string, string> metadata = 8;

  // Nested enum
  enum Gender {
    GENDER_UNSPECIFIED = 0;
    MALE = 1;
    FEMALE = 2;
    NON_BINARY = 3;
  }

  // Nested message
  message PhoneNumber {
    string number = 1;
    PhoneType type = 2;

    // Nested enum inside nested message
    enum PhoneType {
      MOBILE = 0;
      HOME = 1;
      WORK = 2;
    }
  }
}
```


Has a generated class of:

```cpp
// Represents a person with various fields, nested messages, and enums.
class person {
public:
  // Serializes rich.example.Person into a protocol buffer.
  iobuf to_proto() const;
  // Serializes rich.example.Person into proto3 JSON.
  iobuf to_json() const;
  // Serializes rich.example.Person into a protocol buffer, in a way that will not cause stalls for large messages.
  seastar::future<iobuf> to_proto_async() const;
  // Serializes rich.example.Person into proto3 JSON, in a way that will not cause stalls for large messages.
  seastar::future<iobuf> to_json_async() const;
  // Deserializes rich.example.Person from a protocol buffer.
  static void from_proto(iobuf_parser*, person*);
  static person from_proto(iobuf_parser*);
  // Deserializes rich.example.Person from a protocol buffer, in a way that will not cause stalls for large messages.
  static seastar::future<> from_proto_async(iobuf_parser*, person*);
  static seastar::future<person> from_proto_async(iobuf_parser*);
  // Deserializes rich.example.Person from JSON.
  static void from_json(iobuf_parser*, person*);
  static person from_json(iobuf_parser*);
  // Deserializes rich.example.Person from JSON, in a way that will not cause stalls for large messages.
  static seastar::future<> from_json_async(iobuf_parser*, person*);
  static seastar::future<person> from_json_async(iobuf_parser*);
  
  bool operator==(const person&) const = default;  
  // Unique identifier for the person
  int32_t get_id() const { return id_; }
  void set_id(int32_t v) { id_ = v; }
  // Full name
  ss::sstring& get_name() { return name_; }
  const ss::sstring& get_name() const { return name_; }
  void set_name(ss::sstring&& v) { name_ = std::move(v); }
  // Email address
  ss::sstring& get_email() { return email_; }
  const ss::sstring& get_email() const { return email_; }
  void set_email(ss::sstring&& v) { email_ = std::move(v); }
  // List of phone numbers
  chunked_vector<person_phone_number>& get_phones() { return phones_; }
  const chunked_vector<person_phone_number>& get_phones() const { return phones_; }
  void set_phones(chunked_vector<person_phone_number>&& v) { phones_ = std::move(v); }
  // Gender enum
  person_gender get_gender() const { return gender_; }
  void set_gender(person_gender v) { gender_ = v; }
  // Address — oneof home or work
  template<typename Self, typename... Args>
  auto&& visit_address(this Self&& self, Args&&... args) { seastar::visit(std::forward<Self>(self).address_, std::forward<Args>(args)...); }
  bool has_home_address() const { return address_.index() == 1; }
  ss::sstring& get_home_address() { return std::get<1>(address_); }
  const ss::sstring& get_home_address() const { return std::get<1>(address_); }
  void set_home_address(ss::sstring&& v) { address_.emplace<1>(std::move(v)); }
  bool has_work_address() const { return address_.index() == 2; }
  ss::sstring& get_work_address() { return std::get<2>(address_); }
  const ss::sstring& get_work_address() const { return std::get<2>(address_); }
  void set_work_address(ss::sstring&& v) { address_.emplace<2>(std::move(v)); }
  // Metadata (string → string map)
  chunked_hash_map<ss::sstring, ss::sstring>& get_metadata() { return metadata_; }
  const chunked_hash_map<ss::sstring, ss::sstring>& get_metadata() const { return metadata_; }
  void set_metadata(chunked_hash_map<ss::sstring, ss::sstring>&& v) { metadata_ = std::move(v); }

private:
  int32_t id_{};
  ss::sstring name_{};
  ss::sstring email_{};
  chunked_vector<person_phone_number> phones_{};
  person_gender gender_{};
  std::variant<std::monostate, ss::sstring, ss::sstring> address_{};
  chunked_hash_map<ss::sstring, ss::sstring> metadata_{};
};
```

The motivation behind this is to make the Admin API much easier to use. In support of this, protobuf services and RPC methods are supported, generating helper classes that handle all the serialization of RPC requests and responses. The protocol used for the generated protos is [ConnectRPC](https://connectrpc.com/), which is already used extensively by the Console and Cloud teams, the protocol is much simpler than gRPC and doesn't require HTTP2 support. Note that we only support unary RPCs for now, streaming RPCs if needed could have support added for them.

There is not yet an example of defining a new Admin API, but that will be plumbed and demonstrated in a followup PR.

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

dotnwat commented:
👏 👏 👏

dotnwat commented:
this is gonna open up some dope use cases

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--fb53f06f-22ff-4dca-a888-ee99d69f9e57-->
<!--TEST RESULTS-->
<!--68eff6e5-4079-4daf-afff-f8802d96e914-->
<!--TEST RESULTS-->
<!--f91de999-866b-4511-8dc4-92004809501c-->
<!--TEST RESULTS-->
<!--4becae8f-2a60-4bb6-a7e5-8ee5bea968a5-->
<!--TEST RESULTS-->
<!--9c0ac72f-9bbb-47b1-975b-229d82bca582-->
<!--TEST RESULTS-->
<!--58f00132-f258-46f5-80d7-deba38b96480-->
<!--TEST RESULTS-->
<!--a4a8d358-5d0e-4594-af17-7a1a920fe45f-->
<!--TEST RESULTS-->
<!--c0a8593e-3107-454f-afc0-d9bcc051219b-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67986">build#67986</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:---------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RaftAvailabilityTest     | test_one_node_down   |                                                                                                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67986#0197ad26-9dc7-4839-93ff-0b807d2cf5e5 | FLAKY         | 18/21    | upstream reliability is '95.60723514211887'. current run reliability is '85.71428571428571'. drift is 9.89295 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67986#0197ad26-9dc2-4b6d-8ddf-422b0872d432 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS             |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68026">build#68026</a></summary>


| test_class               | test_method   | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:-------------------------|:--------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageScrubberTest | test_scrubber | {"cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68026#0197ae28-a863-4b89-89f4-9a176c9a967c | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68194">build#68194</a></summary>


| test_class              | test_method   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:------------------------|:--------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DisablingPartitionsTest | test_disable  |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68194#0197c1ce-731c-4b54-927f-3d9615d6c6f9 | FLAKY         | 17/21    | upstream reliability is '92.1965317919075'. current run reliability is '80.95238095238095'. drift is 11.24415 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68778">build#68778</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                               |
|:-------------------------|:---------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68778#0197f657-0c62-470d-af1f-95519a010e41 | FLAKY         | 19/21    | upstream reliability is '98.6449864498645'. current run reliability is '90.47619047619048'. drift is 8.1688 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68840">build#68840</a></summary>


| test_class                          | test_method                    | test_arguments             | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:------------------------------------|:-------------------------------|:---------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodePostRestartProbeTest            | post_restart_probe_test        |                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68840#0197fa7a-c35c-4faf-8322-b9e9396d8cdf | FLAKY         | 20/21    | upstream reliability is '99.53917050691244'. current run reliability is '95.23809523809523'. drift is 4.30108 and the allowed drift is set to 50. The test should PASS |
| NodeDecommissionSpaceManagementTest | test_decommission              | {"single_partition": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68840#0197fa7a-c363-41cf-be90-0be1703b1286 | FLAKY         | 19/21    | upstream reliability is '99.07407407407408'. current run reliability is '90.47619047619048'. drift is 8.59788 and the allowed drift is set to 50. The test should PASS |
| RaftAvailabilityTest                | test_controller_node_isolation |                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68840#0197fa7a-c357-4cc9-849c-14a051e9fd61 | FLAKY         | 20/21    | upstream reliability is '98.57142857142858'. current run reliability is '95.23809523809523'. drift is 3.33333 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68925">build#68925</a></summary>


| test_class               | test_method           | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:----------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeThrottlingTest   | test_basic_throttling | {"catalog_type": "nessie", "cloud_storage_type": 1}                                                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68925#01980a0c-1cae-4d19-ab7d-bd49600f3b68 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| RandomNodeOperationsTest | test_node_operations  | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68925#01980a0d-82fc-4d9e-a36f-316b45a955be | FLAKY         | 20/21    | upstream reliability is '98.87218045112782'. current run reliability is '95.23809523809523'. drift is 3.63409 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69006">build#69006</a></summary>


| test_class               | test_method             | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MaintenanceTest          | test_maintenance_sticky | {"use_rpk": false}                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69006#01980e8c-43c9-4e43-9461-c10d49b450ff | FLAKY         | 20/21    | upstream reliability is '98.18181818181819'. current run reliability is '95.23809523809523'. drift is 2.94372 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations    | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69006#01980e8c-43ca-4b9d-9ae2-1c9049be2edb | FLAKY         | 19/21    | upstream reliability is '99.44444444444444'. current run reliability is '90.47619047619048'. drift is 8.96825 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/69119">build#69119</a></summary>


| test_class               | test_method          | test_arguments                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:---------------------|:------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69119#019814d0-ed07-454d-92ac-b3e4e378d9ff | FLAKY         | 19/21    | upstream reliability is '98.40637450199203'. current run reliability is '90.47619047619048'. drift is 7.93018 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": true}          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/69119#019814d0-ed0c-4d0a-b96e-089aed89cc12 | FLAKY         | 20/21    | upstream reliability is '98.39357429718876'. current run reliability is '95.23809523809523'. drift is 3.15548 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

dotnwat commented:
seems like some real coverage gaps in the parser?

<img width="1307" alt="image" src="https://github.com/user-attachments/assets/160f666e-f58a-4758-b8de-0eeb909989e2" />

rockwotj commented:
> in the parser?

What aspect of parsing do you mean? The parser has decent coverage from the iceberg dynamic reading of protos approach. The JSON support could certainly have more coverage. Some of the lines missing coverage are likely because we don't check weird map keys and there are things like NaN, Inf we need to add support for

rockwotj commented:
Let me add our conformance protobuf to the round_trip_test which should bump up coverage

dotnwat commented:
>What aspect of parsing do you mean?

Oh I was mistaken I thought when I had looked earlier that the parser was 100% new in this PR. Indeed it is not.

rockwotj commented:
The parser could have better coverage tho, I'll look at improving that.

rockwotj commented:
I added the conformance proto to the manually serialized code and caught a bunch of issues.

rockwotj commented:
I also removed the non-async versions of parsing/serializing. Maybe we want those back someday, but for now I will remove. The conformance proto was taking a minute to compile locally with all the duplicated code between the async and non async versions.

rockwotj commented:
> seems like some real coverage gaps in the parser?
> <img alt="image" width="1307" src="https://private-user-images.githubusercontent.com/242417/460043956-160f666e-f58a-4758-b8de-0eeb909989e2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTEwNTYzNDYsIm5iZiI6MTc1MTA1NjA0NiwicGF0aCI6Ii8yNDI0MTcvNDYwMDQzOTU2LTE2MGY2NjZlLWY1OGEtNDc1OC1iOGRlLTBlZWI5MDk5ODllMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNjI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDYyN1QyMDI3MjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05ZWEzNWEyZTYzODc3OTlhOTJjYTdiNTViNTUxZTM3Mzk1YzlmMzcwOGExNWU3ZTUyNDljZTQ1YWVkMTJiMmFkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.LudkJGJKLSojdsjgOa38Z2C3jlIU6LFyfmRLjtj2zCk">

Post conformance testing:

<img width="1430" alt="Screenshot 2025-06-27 at 3 27 17 PM" src="https://github.com/user-attachments/assets/2525a2f1-ad9d-4237-a2b6-48cd0f5e9eb9" />

nvartolomei commented on bazel/pbgen/example.proto:
```diff
@@ -0,0 +1,87 @@
+  int32 id = 1;
+
+  // Full name
+  string name = 2 [ctype = CORD];
```
First time I got across this. Didn't get to later commits yet but this seems to be a directive to use iobuf instead of linear string.

We need a short document for writing "seastar" friendly protos.

nvartolomei replied:
I don't see an example with the `bytes` type.

rockwotj replied:
I've removed this style (there are issues with it) and made an examples in the `proto/redpanda/pbgen/options.proto` file.

`bytes` always give you `iobuf` there is no options.

nvartolomei commented on bazel/pbgen/example.proto:
```diff
@@ -0,0 +1,87 @@
+      string currency = 2;
+    }
+
+    Budget budget = 3 [(rp.core.pb.ptr) = true];
```
Need this documented in a short md file at least please. Why and when to use it.

rockwotj replied:
Done in `proto/redpanda/pbgen/options.proto`

nvartolomei commented on bazel/pbgen/main.go:
```diff
@@ -0,0 +1,640 @@
+}
+
+func (b *codewriter) Printf(msg string, args ...any) {
+	_, _ = b.content.WriteString(b.indent + fmt.Sprintf(msg, args...))
```
nit of course but I'd expect 2 calls to WriteString without the concatenation to be more efficient

and why not use fmt.Fprintf too to avoid further allocations/memcpyi

rockwotj replied:
Because I don't really care for optimizing this code, it's basically instant compared to compiling the output and this was the most concise :)

nvartolomei commented on bazel/pbgen/main.go:
```diff
@@ -0,0 +1,640 @@
+			} else {
+				w.Printf("%s& get_%s() { return _%s; }\n", typ, field.Name(), field.Name())
+				w.Printf("const %s& get_%s() const { return _%s; }\n", typ, field.Name(), field.Name())
+				w.Printf("void set_%s(%s v) { _%s = std::move(v) }\n", field.Name(), typ, field.Name())
```
missing semicolons after `std::move(v)`

nvartolomei replied:
Test coverage gap?

rockwotj replied:
Fixed in a latter commit (sorry I did a bad job with commits because I worked on it over a very long time).

nvartolomei commented on bazel/pbgen/main.go:
```diff
@@ -0,0 +1,640 @@
+	for i := range msg.Fields().Len() {
+		field := msg.Fields().Get(i)
+		typ, trivial := g.translateType(field)
+		if oneof := field.ContainingOneof(); oneof != nil {
```
consider extracting this oneOf logic to a method. it's a bit too much for a single method with double nesting, shadowing `i` variable

rockwotj replied:
Good call, will do

rockwotj replied:
done in bd37d7f908

nvartolomei commented on bazel/pbgen/main.go:
```diff
@@ -0,0 +1,640 @@
+			}
+			g.leadingComments(field, w)
+			if trivial {
+				w.Printf("%s get_%s() const { return std::get<%d>(_%s); }\n", typ, field.Name(), idx, oneof.Name())
```
Official Cpp library (and I believe others) return default value if variant is not set. Are we deliberately choosing to throw with bad variant access?

rockwotj replied:
Yes sort of. I have not thought deeply about this, but I think it's never intentional to get the default value. TBH I think originally Google did this because there were no exceptions in their C++ and the other options (crashing or having a more complex accessor method) were just worse. Everything else is really just a port of the C++ implementaiton

nvartolomei commented on bazel/pbgen/main.go:
```diff
@@ -0,0 +1,640 @@
+	defer w.Println("}")
+	w.Indent()
+	defer w.Dedent()
+	// TODO!
```
i expect these to be filled in later commits but `panic("not implemented")` is always nicer than a no-effect `TODO`

can continue reviewing without second thoughts if i see a panic/err return but now i have to push this context on the stack

rockwotj replied:
Yes agreed, I should have panic'd

nvartolomei commented on bazel/pbgen/main.go:
```diff
@@ -549,40 +551,107 @@ func (g *implGenerator) generateRepeatedFieldRead(f protoreflect.FieldDescriptor
+		// TODO: Handle packed fallback of grabbing the latest
+		w.Println("if (tag.wire_type != serde::pb::wire_type::i32) [[unlikely]] {")
+		w.Indent()
+		w.Println(`throw std::runtime_error("unexpected wire type");`)
```
nit: `if (tag != expected) { throw }` is repeated a good couple of times. Can be extracted as a `g.wireTypeGuard("i32")` and the message could be improved by adding info about expected wire type vs actual

rockwotj replied:
I need to just pull this out as a helper in `support.h` and then get rid of the extra codegen

rockwotj replied:
I addressed this in https://github.com/redpanda-data/redpanda/pull/26468/commits/9340f8041c52262b747869f56fcc769230fad97d which greatly simplifies the codegen

nvartolomei commented on bazel/pbgen/tmp/example.pb.cc:
```diff
@@ -1,520 +0,0 @@
-// Code generated by bazel/pbgen. DO NOT EDIT.
```
Can we return generated files back? Help catching (unexpected) changes in the compiler and reviewing changes.

Maybe as Go ([golden](https://github.com/franiglesias/golden)) tests?

rockwotj replied:
Done

rockwotj replied:
https://github.com/redpanda-data/redpanda/pull/26468/commits/84b4ee3eac4eb61114b26f4a33893a392a89919a

nvartolomei commented on bazel/pbgen/main.go:
```diff
@@ -500,8 +508,39 @@ func (g *implGenerator) generateMessageWriteTagInvoke(msg protoreflect.MessageDe
+}
+
+func (g *implGenerator) generateSingularFieldWrite(f protoreflect.FieldDescriptor, w *codewriter, async bool) {
+	panic("unimplemented")
```
👍

nvartolomei reviewed (commented):
Nice.

How far do we expect these generated structs to reach inside redpanda codebase? Will we require in any way that they don't go say past RPC layer? If no, then I'd like to request to consider splitting the struct generation from serde and rpc code (mostly concerned about includes).

Skipped over some parts but overall looks pretty good to merge. Ah, I'd love some golden tests for Golang generator before merging.

rockwotj commented:
> How far do we expect these generated structs to reach inside redpanda codebase? Will we require in any way that they don't go say past RPC layer? If no, then I'd like to request to consider splitting the struct generation from serde and rpc code (mostly concerned about includes).

I've tried to keep includes to be a minimal set depending on what proto features are used. Currently there are no plans to use these outside the admin layer, but that may change longer term. I think we should be splitting proto files so they are minimal. IE "core" proto files should be separate from the RPC only protos.

I believe most people will just cargo cult here, and I'll hopefully setup some good patterns in a followup when we plug this all into admin server.

rockwotj commented:
Okay I've simplified the codegen a great deal and I'm keen to get this in to not let it bit rot.

rockwotj committed:
serde/json/writer: allow appending raw json

This allows writing raw JSON directly to the output stream without
additional parsing or validation. This is useful for embedding
pre-formatted JSON content into larger JSON structures.

rockwotj committed:
serde/json: remove experimental namespace

The parser is fairly well tested, and I don't see a reason to warn usage
with this experimental prefix.

rockwotj committed:
iobuf: make comparable to std::string_view

Make comparing as well as equality between iobuf and string_view
possible. We will use this for binary search in the protoc plugin json
parsing.

rockwotj committed:
serde/protobuf: add RPC service framework

Add RPC service framework for protobuf messages with ConnectRPC
compatibility layer. Provides infrastructure for defining and
implementing RPC services using protobuf definitions.

rockwotj committed:
serde/protobuf: add round-trip testing

Add comprehensive round-trip testing to validate protobuf serialization
and deserialization correctness against the official protobuf library.

rockwotj committed:
pbgen: add protoc plugin for wire format generation

Add a protoc plugin for generating RedPanda-compatible protobuf code.
Includes bazel build rules, plugin framework, and wire format code
generation for native C++ serialization from .proto files.

rockwotj committed:
pbgen: add golden file testing framework

Add golden file testing infrastructure and conformance testing using
official protobuf test messages. Includes bazel rules for generating
and validating golden files against expected outputs.

rockwotj commented:
Cleaned up the commit history if that helps

rockwotj commented:
Force push: reworded some commit messages

mmaslankaprv commented on src/v/serde/json/writer.h:
```diff
@@ -78,6 +78,15 @@ class writer {
+    //
+    // This method is dangerous, as you need to ensure that the JSON is valid.
+    void append_raw_json(iobuf&& b) {
+        append_delimiter();
```
nit: it is confusing tha the comment states it appends raw JSON without delimiters but here we are actually appending delimiter and set next in the last line

rockwotj replied:
fixed

mmaslankaprv commented on src/v/iceberg/conversion/values_json.cc:
```diff
@@ -111,25 +111,25 @@ std::optional<primitive_value> convert_primitive(
 
 ss::future<> decode_list(
-  experimental::serde::json::parser& p,
+  serde::json::parser& p,
```
nit: this should probably land in the commit changing `json` namespace

rockwotj replied:
fixed

rockwotj committed:
serde/protobuf: introduce helpers for JSON serialization

These will be used in the generated code to implement JSON serialization
and deserialization for the generated protobuf messages.

rockwotj committed:
proto: add protobuf definitions and documentation

Add protobuf service definitions, RPC metadata support, and buf.yaml
configuration. Includes documentation on how to write and use protobuf
definitions within Redpanda.

rockwotj committed:
serde/protobuf: add wire format parsing utilities

Add comprehensive wire format parsing library for protobuf messages.
Includes support for all protobuf wire types, varint encoding/decoding,
and utilities for parsing protobuf fields from binary data.
